crates/ruff/src/rules/flake8_pyi/rules/fixed_return_type.rs

```diff
@@ -0,0 +1,250 @@
+pub struct FixedReturnType {
+    method_name: String,
+}
+
+impl Violation for FixedReturnType {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let FixedReturnType { method_name } = self;
+        format!("{method_name} methods on classes usually return `self` at runtime. Consider using `typing_extensions.Self` as return type")
+    }
+}
+
+const SELF_RETURNING_METHODS: &[&str] = &[
+    "__iadd__",
+    "__isub__",
+    "__ior__",
+    "__iand__",
+    "__imul__",
+    "__itruediv__",
+    "__ifloordiv__",
+    "__new__",
+    "__enter__",
+    "__aenter__",
+];
+
+const ITERATOR_BASES: &[&[&str]] = &[
+    &["typing", "Iterator"],
+    &["typing", "AsyncIterator"],
+    &["collections", "abc", "Iterator"],
+    &["collections", "abc", "AsyncIterator"],
+];
+
+const ASYNC_ITER_RETURN_TYPES: &[&[&str]] = &[
+    &["typing", "AsyncIterable"],
+    &["typing", "AsyncIterator"],
+    &["collections", "abc", "AsyncIterable"],
+    &["collections", "abc", "AsyncIterator"],
+];
+
+const SYNC_ITER_RETURN_TYPES: &[&[&str]] = &[
+    &["typing", "Iterable"],
+    &["typing", "Iterator"],
+    &["collections", "abc", "Iterable"],
+    &["collections", "abc", "Iterator"],
+];
+
+/// PYI034
+pub(crate) fn fixed_return_type(
+    checker: &mut Checker,
+    bases: &[Expr],
+    body: &[Stmt],
+    class_decorator_list: &[Expr],
+) {
+    // If class is final, skip
+    if class_decorator_list.iter().any(|expr| {
+        checker
+            .semantic_model()
+            .match_typing_expr(map_callable(expr), "final")
+    }) {
+        return;
+    }
+
+    let mut is_iter_subclass: bool = false;
+    if bases.len() == 1 {
+        let base = if let Expr::Subscript(ast::ExprSubscript { value, .. }) = &bases[0] {
+            // Ex) class Foo(Iterator[T]):
+            value
+        } else {
+            // Ex) class Foo(Iterator):
+            &bases[0]
+        };
+        is_iter_subclass = checker
+            .semantic_model()
+            .resolve_call_path(base)
+            .map_or(false, |call_path| {
+                ITERATOR_BASES.contains(&call_path.as_slice())
+            });
+    }
+
+    for stmt in body {
+        let (
+            Stmt::FunctionDef(ast::StmtFunctionDef {
+                decorator_list: func_decorator_list,
+                name: method_name,
+                returns,
+                ..
+            }) | Stmt::AsyncFunctionDef(ast::StmtAsyncFunctionDef {
+                decorator_list: func_decorator_list,
+                name: method_name,
+                returns,
+                ..
+            })
+        ) = stmt else {
+            continue;
+        };
+
+        // Skip abstractmethods and overloaded methods
+        if is_abstract(checker.semantic_model(), func_decorator_list)
+            || is_overload(checker.semantic_model(), func_decorator_list)
+        {
+            continue;
+        }
+
+        let Some(returns) = returns else {
+            continue;
+        };
```

Comment:
Nit: It could be worth to move this above line 177 because this is a very cheap check compared to performing any operations on the semantic model.